### PR TITLE
SIMD.js load/store type-spec and more ops enabling

### PR DESCRIPTION
**Pull request by nmostafa**
- Enabled type-spec of more bitwise/arithmetic ops in Int32x4 and Float32x4
- Enabled shuffle/swizzle type-spec with arbitrary lane indices (not stricted to x86 SHUF* ops requirements). Enabled lowerer.
- Type-spec of Float32x4/Int32x4 load/store with bound check hoisting and elimination.
- Asm.js tests were not running with TypeSpec (even with -simd128typespec), since asm.js was enabled by default. Fixed this by additing "-asmjs-" to those tests.

[Review on Reviewable](https://reviewable.io/reviews/microsoft/chakracore/113)

### CODE REVIEW

**Comment by nmostafa** [comment]
@LouisLaf, can you look at the globOpt part ?
@MikeHolman , can you look at the Lowerer part ?

**Comment by dilijev** [comment]
@nmostafa looks like a test file is missing a copyright

```
16:16:30 --------------
16:16:30 --- ERRORS ---
16:16:30 test/SIMD.TypeSpec/loop-test-2.js ... does not contain a correct Microsoft copyright notice.
16:16:30 --------------
```

**Comment by LouisLaf** [comment]
Globopt code looks good.

**Comment by MikeHolman** [comment]
Reviewed 6 of 34 files at r1.
Review status: 6 of 34 files reviewed at latest revision, 7 unresolved discussions.

---

*[lib/Backend/Lower.cpp, line 12960 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8Wf-yBMVWoJt481TVL:-K8W587AN7L8r-06kUwp:-1075757189) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/Lower.cpp#L12960)):*
Why remove const?

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 830 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8WEV9LBc-4HqcFME2c:-K8WEV9M4c2fhR_QhGtr:1620346732) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L830)):*
should be combined into previous if

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 830 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8WPdFRakuH6EYY9KNh:-K8WPdFSIQNXmp2mbRKp:1872157677) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L830)):*
should we use pshufd for swizzle_i4?

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 837 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8WJP1tsAjEJNMTLEEr:-K8WJP1tsAjEJNMTLEEs:114126102) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L837)):*
nit: maybe it makes no difference (you would know better), but maybe better to movapd for swizzle_d2 and movdqa for swizzle_i4?

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 946 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8WbQ3uu1A6vhOn851P:-K8WbQ3uu1A6vhOn851Q:-256098818) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L946)):*
why not do this at compile time and have a different mask depending on which is minority lane? seems better than runtime shift every time. (also, comment is wrong... was confusing me for a bit why you did shift on temp2)

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 964 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8WSGnScXQqa5Evs6In:-K8WSGnScXQqa5Evs6Io:1629748593) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L964)):*
basically same comment as above, we are using float simd ops even for ints here

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 1576 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8Wcv3CsRdTb5X0SuY6:-K8Wcv3CsRdTb5X0SuY7:-1513438623) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L1576)):*
and again, would be good to use correct type depending if it is int or float

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/113)*

File: lib/Backend/LowerMDSharedSimd128.cpp [hunk 1]
```diff
@@ -1042,88 +1155,178 @@ IR::Instr* LowererMD::Simd128LowerStoreElem(IR::Instr *instr)
...
+    IR::ArrayRegOpnd *arrayRegOpnd = indirOpnd->GetBaseOpnd()->AsArrayRegOpnd();
+    IR::RegOpnd *headSegmentOpnd;
+    
+    if (arrayRegOpnd->HeadSegmentSym())
+    {
+        headSegmentOpnd = IR::RegOpnd::New(arrayRegOpnd->HeadSegmentSym(), TyMachPtr, m_func);
+    }
+    else
+    {
+        // REVIEW: Is this needed ? Shouldn't globOpt make sure headSegSym is set and alive ?
```
Comment (MikeHolman) [inline lib/Backend/LowerMDSharedSimd128.cpp]: @LouisLaf maybe you want to help with this question?

Comment (LouisLaf) [inline lib/Backend/LowerMDSharedSimd128.cpp]: An ArrayRegOpnd for a LdElem or StElam should always have a headSegSym.  It just doesn't create it for a LdLen.

**Comment by nmostafa** [comment]
Review status: 6 of 34 files reviewed at latest revision, 6 unresolved discussions.

---

*[lib/Backend/Lower.cpp, line 12960 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8Wf-yBMVWoJt481TVL:-K8Wpg1jN1mBm3ncyWQn:-475205071) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/Lower.cpp#L12960)):*
Changed method to static member.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 837 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8WJP1tsAjEJNMTLEEr:-K8eMXvJaLntAO9CsRbp:-1887353162) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L837)):*
Currently the encoder only support MOVAPS/MOVUPS for SSE2 moves. I doubt this will be a perf issue, but we can revisit (and add support for instr), if it shows up.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 946 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8WbQ3uu1A6vhOn851P:-K8eNoDMWLGXJxkcUmKw:-1278773285) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L946)):*
Yes, we can. Will change it.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 964 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8WSGnScXQqa5Evs6In:-K8eMwbYQFqx-NCr66tB:1801257787) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L964)):*
Again, some ops are missing in the encoder (e.g. PANDN), will add support if it shows up as a perf issue. 

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 1316 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8VgHtTR3_cSEgTPIgs-r2-1316:-K8eNN3hRm9tpfhpBGxm:-480976131) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L1316)):*
ok, will replace with Assert(UNREACHED)

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 1576 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8Wcv3CsRdTb5X0SuY6:-K8eNQxtEo0j1AKhvGpH:-709678398) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L1576)):*
Will fix.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/113)*

**Comment by nmostafa** [comment]
*[lib/Backend/LowerMDSharedSimd128.cpp, line 1316 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8VgHtTR3_cSEgTPIgs-r2-1316:-K8e_5ydzZwgbvr5Entw:1793366349) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L1316)):*
Apparently , head segment is only loaded in GlobOpt for generic arrays only (non typed arrays) in OptArraySrc().
```   
       const bool isLikelyJsArray = !baseValueType.IsLikelyTypedArray();
        ....
        if(doHeadSegmentLoad && isLikelyJsArray)
        {
            // For javascript arrays, the head segment is required to load the head segment length
            InsertHeadSegmentLoad();
        }
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/113)*

**Comment by nmostafa** [comment]
*[lib/Backend/LowerMDSharedSimd128.cpp, line 1576 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/113#-K8Wcv3CsRdTb5X0SuY6:-K8edF70hVLPmZEch3Yx:1932752556) ([raw file](https://github.com/microsoft/chakracore/blob/7613e267a2e6fff2d4565e10f1f73d2ac150cfee/lib/Backend/LowerMDSharedSimd128.cpp#L1576)):*
Actually, there is a subtle difference between SHUFPS and PSHUFD. SHUFPS shuffles from dst and src, while PSHUFD shuffles from src only. We need SHUFPS here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/113)*

**Comment by MikeHolman** [comment]
lgtm

**Comment by dilijev** [comment]
@dotnet-bot test EOL Check please

**Comment by mmitche** [comment]
@dotnet-bot test EOL Check please
